### PR TITLE
Add enrichment models - Recognize Anything, Tag2text, Grounding DINO and Segment Anything Model

### PR DESCRIPTION
**Pull request by dnth**
This PR adds zero shot models such as [Recognize Anything](https://github.com/xinyu1205/recognize-anything), [Tag2text](https://github.com/xinyu1205/recognize-anything), [Grounding DINO](https://github.com/IDEA-Research/GroundingDINO) and [Segment Anything](https://github.com/facebookresearch/segment-anything) models as an enrichment option for fastdup.

Here's a Colab notebook to run the following demo - https://colab.research.google.com/drive/1vTac7oTFIZ2eL1v-CDlSJTuCvbi3j0CB?usp=sharing

The enrichment API usage is as follows:

```python
import fastdup
fd = fastdup.create(input_dir='./coco_minitrain_25k')
fd.run()

# Enrich data with labels from Recognize Anything Model (RAM)
df = fd.enrich(task='zero-shot-classification', model='recognize-anything-model', num_rows=5)
```

Users can continue adding enrichment with other models by repeatedly calling the `.enrich` method and specifying an input dataframe. 

```python
# Enrich data with Grounding DINO model.
df = fd.enrich(task='zero-shot-detection', model='grounding-dino', input_df=df)
```

The dataframe will contain the enriched data

image

The following is a plot using the labels generated by the models

image

User can continue to enrich data with Segment Anything model to generate masks from the Grounding DINO bounding boxes.

```python
df = fd.enrich(task='zero-shot-segmentation', model='segment-anything', input_df=df)
```
The results are as follows

image

### CODE REVIEW

File: fastdup/models/grounding_dino.py [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+
+    grounding_dino_transform = T.Compose(
+        [
+            T.RandomResize([800], max_size=1333),
+            T.ToTensor(),
+            T.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225]),
+        ]
+    )
+except ImportError as e:
+    print(e)
```
Comment (dbickson) [inline fastdup/models/grounding_dino.py:23]: @dnth please use fastdup_sentry_capture_exception to report any exception (vs. print(e))

File: fastdup/models/grounding_dino.py [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+
+    def _download_file(self, target_path, download_url):
+        """Helper function to download a file if it doesn't exist."""
+        if not os.path.exists(target_path):
+            logger.info(
+                f"File not found at {target_path}. Downloading from {download_url}..."
+            )
+            try:
+                subprocess.run(
+                    ["wget", "-O", target_path, download_url],
```
Comment (dbickson) [inline fastdup/models/grounding_dino.py:56]: @dnth please use fastdup.utils.download_from_web instead, wget is supported only on linux but not on mac, on mac it is curl

File: fastdup/models/sam.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+            )
+            subprocess.run(["wget", "-O", self.weights, download_url])
+            logger.info("Model downloaded.")
+        build_sam = sam_model_registry[self.vit_type]
+        sam_model = SamPredictor(build_sam(checkpoint=self.weights))
+        sam_model.mode = sam_model.model.to(self.device)
+        return sam_model
+
+    def run_inference(self, image_path, bboxes):
+        image = cv2.imread(image_path)
```
Comment (dbickson) [inline fastdup/models/sam.py:51]: please use utils.fastdup_imread instead as it supports also heic, heif, 16 bit grayscale

File: fastdup/models/tag2text.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        self.device = device
+        self.model = self._load_model()
+
+    def _load_model(self) -> torch.nn.Module:
+        if not os.path.exists(self.model_path):
+            download_url = "https://huggingface.co/spaces/xinyu1205/Recognize_Anything-Tag2Text/resolve/main/tag2text_swin_14m.pth"
+            logger.info(
+                f"Model file not found. Downloading model from {download_url}..."
+            )
+            subprocess.run(
```
Comment (dbickson) [inline fastdup/models/tag2text.py:42]: same comments about download

File: fastdup/models/ram.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        with contextlib.redirect_stdout(io.StringIO()):
+            model = ram(pretrained=self.model_path, image_size=384, vit="swin_l")
+            model.eval()
+            model = model.to(self.device)
+        logger.info(f"Model loaded to device - {self.device}")
+        return model
+
+    def run_inference(self, image_path: str) -> str:
+        transform = get_transform(image_size=384)
+        image = transform(Image.open(image_path)).unsqueeze(0).to(self.device)
```
Comment (dbickson) [inline fastdup/models/ram.py:61]: please use fastdup_imread and make sure the image is in the format you need

File: fastdup/models/sam.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
+try:
+    import torch
+except ImportError:
+    raise ImportError(
+        "The `torch` package is not installed. Please run `pip install torch` or equivalent."
```
Comment (dbickson) [inline fastdup/models/sam.py:5]: use fastdup_sentry_report_exception

**Comment by dbickson** [comment]
@dnth very nice!
Two ninor comments.
1) Why read fastdup filenames from folder, when you can run fd.annotations() and get a dataframe with all the filenames?
2) the output of the bounding boxes should be in fastdup consumable format namely columns of filename, row_y, col_x, width, hight => one per row, this will allow running fastdup again on the annotated bb for further analysis without any conversions

**Comment by dnth** [comment]
> @dnth very nice! Two ninor comments.
> 
> 1. Why read fastdup filenames from folder, when you can run fd.annotations() and get a dataframe with all the filenames?
> 2. the output of the bounding boxes should be in fastdup consumable format namely columns of filename, row_y, col_x, width, hight => one per row, this will allow running fastdup again on the annotated bb for further analysis without any conversions

Hi @dbickson as a user, I would like to run the enrichment first and then run fastdup on top of the enriched labels. That is the reason why I read the filenames from a folder.

I imagine the flow would look like

```python
import fastdup
from fastdup.utils import get_images_from_path
import pandas as pd

# Set path and filenames
fd = fastdup.create(input_dir='./coco_minitrain_25k')
filenames = get_images_from_path(fd.input_dir)
df = pd.DataFrame(filenames, columns=["filenames"])

# Enrich
df = fd.enrich(task='zero-shot-classification', model='recognize-anything-model', num_rows=5, input_df=df)
```

This currently throws a runtime error because fastdup expects users to run first.

```shell
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Cell In[4], line 1
----> 1 df = fd.enrich(task='zero-shot-classification', model='recognize-anything-model', num_rows=5)

File ~/anaconda3/envs/mmdet-integration/lib/python3.10/site-packages/fastdup/fastdup_controller.py:1333, in FastdupController.enrich(self, task, model, input_df, user_tags, num_rows, input_col, output_col, device)
   1330 def enrich(self, task, model, input_df=None, user_tags="None", num_rows=None, input_col=None, output_col=None, device=None):
   1332     if input_df is None:
-> 1333         df = self.annotations(valid_only=True)
   1334         df.drop(["index", "error_code", "is_valid", "fd_index"], axis=1, inplace=True)
   1335     else:

File ~/anaconda3/envs/mmdet-integration/lib/python3.10/site-packages/fastdup/fastdup_controller.py:197, in FastdupController.annotations(self, valid_only)
    191 """
    192 Get annotation as data frame
    193 :param valid_only: if True, return only valid annotations
    194 :return: pandas dataframe
    195 """
    196 if not self._fastdup_applied:
--> 197     raise RuntimeError('Fastdup was not applied yet, call run() first')
    198 df_annot = self._df_annot.query(f'{FD.ANNOT_VALID}') if valid_only and self._df_annot is not None \
    199     else self._df_annot
    200 return df_annot

RuntimeError: Fastdup was not applied yet, call run() first
```

The two options currently are: 
+ A - Load data -> Run fastdup -> Enrich -> Run fastdup again
+ B - Load data -> Enrich -> Run fastdup

I would prefer option B. What do you think?

**Comment by dnth** [comment]
@dbickson I've added a utility function which can can convert the bounding boxes into json and to be consumed with fastdup.

Usage
```python
from fastdup.models.utils import convert_to_coco_format
convert_to_coco_format(df, bbox_col='grounding_dino_bboxes', label_col='grounding_dino_labels', json_filename='grounding_dino_annot_coco_format.json')
```

After that, users can conveniently use them in fastdup

```python
import fastdup
fd = fastdup.create(input_dir="./")
fd.run(annotations="grounding_dino_annot_coco_format.json", overwrite=True)
```
Users can get the fastdup annotations format by running `fd.annotations()`
image


The above Colab notebook is now updated with all changes.

**Comment by dnth** [comment]
Additionally, I've also included the plotting function in `fastdup.models.utils`

```python
from fastdup.models.utils import plot_annotations
plot_annotations(df, image_col='filename', tags_col='ram_tags', bbox_col='grounding_dino_bboxes', scores_col='grounding_dino_scores', labels_col='grounding_dino_labels', num_rows=10)
```
image

```python
plot_annotations(df, image_col='filename', tags_col='ram_tags', bbox_col='grounding_dino_bboxes', scores_col='grounding_dino_scores', labels_col='grounding_dino_labels', masks_col='sam_masks', num_rows=5)
```
image

**Comment by dnth** [comment]
Merged manually into version 1.46

**Pull request closed without merging**